torch/csrc/utils/tensor_new.cpp

```diff
@@ -506,18 +499,25 @@ Tensor tensor_ctor(const Type& type, PyObject* args, PyObject* kwargs) {
   ParsedArgs<4> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
   if (r.idx == 0) {
+    PyObject* data = r.pyobject(0);
+    if (THPVariable_Check(data)) {
+      PyErr_WarnEx(PyExc_UserWarning,
```

Comment:
I agree with @gchanan 's comment on last PR that this should mention `as_tensor`.

Reply:
I see, thanks a lot for catching this! So how should I do this correctly, can I:
1. raise similar warnings for `as_tensor` when `type_inference` is false (copy construct)
2. no need to detach the new tensor

Reply:
This is a good question.  My opinion is that we don't need to go through the whole warning song and dance with `as_tensor` because it doesn't have a `requires_grad` argument (precisely for the reason mentioned).  We just need to add to the `as_tensor` document that it either returns the tensor passed in or it returns a new Tensor with `requires_grad = False` (it follows that you don't need to call detach and detach isn't preferred from the fact that `requires_grad = False`).

Reply:
Sounds great! Thanks for the clarifications! Indeed `as_tensor` doesn't have a `requires_grad` in args and so no need any detach or warning in the code path. I will add a note to the docs accordingly.

Reply:
I'd say this is confusing. Shouldn't `torch.as_tensor(tensor)` be differentiable?

Reply:
Yes, I think it should be differentiable.

Reply:
Maybe. I'd argue that if users do `as_tensor(something)`, they usually don't know (or didn't bother to figure out) if `something` is a tensor or not, in which case they certainly wouldn't be thinking to themselves "hmm the gradients will glowing back through here, and then ....".

Reply:
@apaszke @fmassa yes, for sure `torch.as_tensor(tensor)` is differentiable. I was wonder if `torch.as_tensor(tensor, dtype=dtype, device=device)`, where dtype and device are different to tensor's, should be differentiable?

Reply:
` yes, for sure torch.as_tensor(tensor) is differentiable.`

Well, that's clearly true because it's the same tensor.

The behavior of `torch.as_tensor(tensor, dtype=dtype, device=device` is a good question.  I guess my thought process is you'd call `as_tensor` on heterogeneous inputs (i.e. some things are tensors, some are lists, etc).  So as @SsnL says, you probably don't care about gradients flowing back through.  If you did, you'd call `to`.  It also seems a little strange that `torch.tensor` isn't differentiable, but `torch.as_tensor` is.  I bet I could be convinced of the opposite though! @apaszke ? @fmassa ?

Reply:
@fmassa Commented on the workchat, I am reposting it here to drive the discussion:

one way of writing code that works for both tensors and numbers / numpy arrays is to always put a as_tensor right in the beginning of the function. So I could see this as a very common pattern:
```
def exp(x):
    x = torch.as_tensor(x)
    return x.exp()
```
I btw think that this is something that we should consider doing in our functions, so that `torch.exp(1.0)` works.

Thus, I would argue that `torch.as_tensor(x)` should not break the computation graph if x is a tensor that requires grad

Reply:
Thanks for reposting my message @weiyangfb !
So, to expand a bit more on the previous message, if we consider the case I mentioned (supporting both numbers and tensors / lists in torch functions), then we would want `torch.as_tensor` to not break the computation graph. Thus, `torch.as_tensor(tensor, dtype=dtype, device=device)` should work as `.to(dtype, device)`, for tensors.

Reply:
Yeah I think the primary use-case for `as_tensor` is writing out functions that can work both with tensors and with regular Python objects, and there's no reason to make them non-differentiable if the inputs already were tensors. I really think it should preserve the graph, and use `.to` to casting if necessary.

Reply:
@apaszke @gchanan @fmassa Thanks a lot for the very helpful insights! I will keep the differentiable behavior of `torch.as_tensor()` as documented.